File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/AlertHelpers.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * See the NOTICE file distributed with this work for additional
 * information regarding copyright ownership.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.appium.espressoserver.lib.helpers;

import android.support.annotation.Nullable;
import android.support.test.uiautomator.By;
import android.support.test.uiautomator.UiObject2;
import android.util.Log;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.regex.Pattern;

import io.appium.espressoserver.lib.handlers.exceptions.InvalidElementStateException;
import io.appium.espressoserver.lib.handlers.exceptions.NoAlertOpenException;

import static android.text.TextUtils.join;
import static io.appium.espressoserver.lib.helpers.AndroidLogger.logger;
import static io.appium.espressoserver.lib.helpers.InteractionHelper.getUiDevice;

public class AlertHelpers {
    private static final String TAG = AlertHelpers.class.getSimpleName();

    private static final String regularAlertButtonResIdPrefix = "android:id/button";
    private static final Pattern regularAlertButtonResIdPattern =
            Pattern.compile("^" + regularAlertButtonResIdPrefix + "\\d+$");
    private static final String alertContentResId = "android:id/content";
    private static final Pattern regularAlertTitleResIdPattern =
            Pattern.compile(".+:id/(alertTitle|custom)$");
    private static final Pattern permissionAlertTitleResIdPattern =
            Pattern.compile(".+:id/permission_message$");
    private static final Pattern permissionAlertButtonResIdPattern =
            Pattern.compile(".+:id/permission_\\w+_button$");
    private static final Pattern alertElementsResIdPattern = Pattern.compile(".+:id/.+");

    private static String buttonResIdByIdx(int index) {
        return String.format("%s%s", regularAlertButtonResIdPrefix, index);
    }

    private static AlertType getAlertType() throws NoAlertOpenException {
        getUiDevice().waitForIdle();

        if (!getUiDevice().findObjects(By.res(regularAlertTitleResIdPattern)).isEmpty()) {
            Log.d(TAG, "Regular alert has been detected");
            return AlertType.REGULAR;
        }
        if (!getUiDevice().findObjects(By.res(permissionAlertTitleResIdPattern)).isEmpty()) {
            Log.d(TAG, "Permission alert has been detected");
            return AlertType.PERMISSION;
        }

        throw new NoAlertOpenException();
    }

    @Nullable
    private static UiObject2 filterButtonByLabel(Collection<UiObject2> buttons, String label) {
        for (UiObject2 button : buttons) {
            if (Objects.equals(button.getText(), label)) {
                return button;
            }
        }
        return null;
    }

    @Nullable
    private static UiObject2 getRegularAlertButton(AlertAction action, @Nullable String buttonLabel) {
        final Map<String, UiObject2> alertButtonsMapping = new HashMap<>();
        final List<Integer> buttonIndexes = new ArrayList<>();
        for (final UiObject2 button : getUiDevice().findObjects(By.res(regularAlertButtonResIdPattern))) {
            final String resId = button.getResourceName();
            alertButtonsMapping.put(resId, button);
            buttonIndexes.add(Integer.parseInt(resId.substring(regularAlertButtonResIdPrefix.length())));
        }
        if (alertButtonsMapping.isEmpty()) {
            return null;
        }
        Log.d(TAG, String.format("Found %d buttons on the alert", alertButtonsMapping.size()));

        if (buttonLabel == null) {
            final int minIdx = Collections.min(buttonIndexes);
            return action == AlertAction.ACCEPT
                    ? alertButtonsMapping.get(buttonResIdByIdx(minIdx))
                    : alertButtonsMapping.get(buttonResIdByIdx(alertButtonsMapping.size() > 1 ? minIdx + 1 : minIdx));
        }
        return filterButtonByLabel(alertButtonsMapping.values(), buttonLabel);
    }

    @Nullable
    private static UiObject2 getPermissionAlertButton(AlertAction action, @Nullable String buttonLabel) {
        final List<UiObject2> buttons = getUiDevice().findObjects(By.res(permissionAlertButtonResIdPattern));
        if (buttons.isEmpty()) {
            return null;
        }
        Log.d(TAG, String.format("Found %d buttons on the alert", buttons.size()));

        if (buttonLabel == null) {
            if (action == AlertAction.ACCEPT) {
                return buttons.size() > 1 ? buttons.get(1) : buttons.get(0);
            }
            if (action == AlertAction.DISMISS) {
                return buttons.get(0);
            }
        } else {
            return filterButtonByLabel(buttons, buttonLabel);
        }
        return null;
    }

    /**
     * Accept or dismiss on-screen alert.
     *
     * @param action      either ACCEPT or DISMISS
     * @param buttonLabel if this parameter is set then the method
     *                    will look for the dialog button with this particular
     *                    text instead of the default one (usually it is the first button
     *                    for ACCEPT and the last one for DISMISS action)
     * @return the actual label of the clicked button
     * @throws NoAlertOpenException         if no dialog is present on the screen
     * @throws InvalidElementStateException if no matching button can be found
     */
    public static String handle(AlertAction action, @Nullable String buttonLabel)
            throws InvalidElementStateException, NoAlertOpenException {
        final AlertType alertType = getAlertType();

        final UiObject2 dstButton = alertType == AlertType.REGULAR
                ? getRegularAlertButton(action, buttonLabel)
                : getPermissionAlertButton(action, buttonLabel);
        if (dstButton == null) {
            throw new InvalidElementStateException("The expected button cannot be detected on the alert");
        }

        final String actualLabel = dstButton.getText();
        logger.info(String.format("Clicking alert button '%s' in order to %s it",
                actualLabel, action.name().toLowerCase()));
        dstButton.click();
        return actualLabel;
    }

    /**
     * @return The actual text of the on-screen dialog. An empty
     * string is going to be returned if the dialog contains no text.
     * @throws NoAlertOpenException if no dialog is present on the screen
     */
    public static String getText() throws NoAlertOpenException {
        final AlertType alertType = getAlertType();

        final List<UiObject2> alertRoots = getUiDevice().findObjects(By.res(alertContentResId));
        if (alertRoots.isEmpty()) {
            Log.w(TAG, "Alert content container is missing");
            throw new NoAlertOpenException();
        }

        final List<String> result = new ArrayList<>();
        final List<UiObject2> alertElements = alertRoots.get(0).findObjects(By.res(alertElementsResIdPattern));
        Log.d(TAG, String.format("Detected %d alert elements", alertElements.size()));
        final String alertButtonsResIdPattern = alertType == AlertType.REGULAR
                ? regularAlertButtonResIdPattern.toString()
                : permissionAlertButtonResIdPattern.toString();
        for (final UiObject2 element : alertElements) {
            final String resName = element.getResourceName();
            if (resName == null || resName.matches(alertButtonsResIdPattern)) {
                continue;
            }

            final String text = element.getText();
            if (text == null || text.isEmpty()) {
                continue;
            }

            result.add(text);
        }
        return join("\n", result);
    }

    public enum AlertAction {
        ACCEPT, DISMISS
    }

    public enum AlertType {
        REGULAR, PERMISSION
    }
}

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/AlertHelpers.kt
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * See the NOTICE file distributed with this work for additional
 * information regarding copyright ownership.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.appium.espressoserver.lib.helpers

import java.util.ArrayList
import java.util.Collections
import java.util.HashMap
import java.util.regex.Pattern
import androidx.test.uiautomator.By
import androidx.test.uiautomator.UiObject2
import io.appium.espressoserver.lib.handlers.exceptions.InvalidElementStateException
import io.appium.espressoserver.lib.handlers.exceptions.NoAlertOpenException

import io.appium.espressoserver.lib.helpers.InteractionHelper.getUiDevice

object AlertHelpers {
    private const val regularAlertButtonResIdPrefix = "android:id/button"
    private val regularAlertButtonResIdPattern = Pattern.compile("^$regularAlertButtonResIdPrefix\\d+$")
    private const val alertContentResId = "android:id/content"
    private val regularAlertTitleResIdPattern = Pattern.compile(".+:id/(alertTitle|custom)$")
    private val permissionAlertTitleResIdPattern = Pattern.compile(".+:id/permission_message$")
    private val permissionAlertButtonResIdPattern = Pattern.compile(".+:id/permission_\\w+_button$")
    private val alertElementsResIdPattern = Pattern.compile(".+:id/.+")

    private val alertType: AlertType
        @Throws(NoAlertOpenException::class)
        get() {
            getUiDevice().waitForIdle()

            if (getUiDevice().findObjects(By.res(regularAlertTitleResIdPattern)).isNotEmpty()) {
                AndroidLogger.debug("Regular alert has been detected")
                return AlertType.REGULAR
            }
            if (getUiDevice().findObjects(By.res(permissionAlertTitleResIdPattern)).isNotEmpty()) {
                AndroidLogger.debug("Permission alert has been detected")
                return AlertType.PERMISSION
            }

            throw NoAlertOpenException()
        }

    /**
     * @return The actual text of the on-screen dialog. An empty
     * string is going to be returned if the dialog contains no text.
     * @throws NoAlertOpenException if no dialog is present on the screen
     */
    val text: String
        @Throws(NoAlertOpenException::class)
        get() {
            val alertType = alertType

            val alertRoots = getUiDevice().findObjects(By.res(alertContentResId))
            if (alertRoots.isEmpty()) {
                AndroidLogger.warn("Alert content container is missing")
                throw NoAlertOpenException()
            }

            val alertElements = alertRoots[0].findObjects(By.res(alertElementsResIdPattern))
            AndroidLogger.debug("Detected ${alertElements.size} alert elements")
            val alertButtonsResIdPattern = if (alertType == AlertType.REGULAR)
                regularAlertButtonResIdPattern.toString()
            else
                permissionAlertButtonResIdPattern.toString()
            return alertElements.toTypedArray()
                    .filter { it.resourceName != null && !it.resourceName.matches(alertButtonsResIdPattern.toRegex()) }
                    .filter { (it.text ?: "").isNotEmpty() }
                    .joinToString(separator = "\n")
        }

    private fun buttonResIdByIdx(index: Int): String {
        return "$regularAlertButtonResIdPrefix$index"
    }

    private fun filterButtonByLabel(buttons: Collection<UiObject2>, label: String?): UiObject2? {
        return buttons.toTypedArray().firstOrNull { it.text == label }
    }

    private fun getRegularAlertButton(action: AlertAction, buttonLabel: String?): UiObject2? {
        val alertButtonsMapping = HashMap<String, UiObject2>()
        val buttonIndexes = ArrayList<Int>()
        for (button in getUiDevice().findObjects(By.res(regularAlertButtonResIdPattern))) {
            val resId = button.resourceName
            alertButtonsMapping[resId] = button
            buttonIndexes.add(Integer.parseInt(resId.substring(regularAlertButtonResIdPrefix.length)))
        }
        if (alertButtonsMapping.isEmpty()) {
            return null
        }
        AndroidLogger.debug("Found ${alertButtonsMapping.size} buttons on the alert")

        if (buttonLabel == null) {
            val minIdx = Collections.min(buttonIndexes)
            return if (action == AlertAction.ACCEPT)
                alertButtonsMapping[buttonResIdByIdx(minIdx)]
            else
                alertButtonsMapping[buttonResIdByIdx(if (alertButtonsMapping.size > 1) minIdx + 1 else minIdx)]
        }
        return filterButtonByLabel(alertButtonsMapping.values, buttonLabel)
    }

    private fun getPermissionAlertButton(action: AlertAction, buttonLabel: String?): UiObject2? {
        val buttons = getUiDevice().findObjects(By.res(permissionAlertButtonResIdPattern))
        if (buttons.isEmpty()) {
            return null
        }
        AndroidLogger.debug("Found ${buttons.size} buttons on the alert")

        if (buttonLabel == null) {
            if (action == AlertAction.ACCEPT) {
                return if (buttons.size > 1) buttons[1] else buttons[0]
            }
            if (action == AlertAction.DISMISS) {
                return buttons[0]
            }
        } else {
            return filterButtonByLabel(buttons, buttonLabel)
        }
        return null
    }

    /**
     * Accept or dismiss on-screen alert.
     *
     * @param action      either ACCEPT or DISMISS
     * @param buttonLabel if this parameter is set then the method
     * will look for the dialog button with this particular
     * text instead of the default one (usually it is the first button
     * for ACCEPT and the last one for DISMISS action)
     * @return the actual label of the clicked button
     * @throws NoAlertOpenException         if no dialog is present on the screen
     * @throws InvalidElementStateException if no matching button can be found
     */
    @Throws(InvalidElementStateException::class, NoAlertOpenException::class)
    fun handle(action: AlertAction, buttonLabel: String?): String {
        val alertType = alertType

        val dstButton = (if (alertType == AlertType.REGULAR)
            getRegularAlertButton(action, buttonLabel)
        else
            getPermissionAlertButton(action, buttonLabel))
                ?: throw InvalidElementStateException("The expected button cannot be detected on the alert")

        val actualLabel = dstButton.text
        AndroidLogger.info("Clicking alert button '$actualLabel' in order to ${action.name.toLowerCase()} it")
        dstButton.click()
        return actualLabel
    }

    enum class AlertAction {
        ACCEPT, DISMISS
    }

    enum class AlertType {
        REGULAR, PERMISSION
    }
}
